pkg/logql/log/labels.go

```diff
@@ -333,34 +458,67 @@ func (b *LabelsBuilder) Map() map[string]string {
 // No grouping is applied and the cache is used when possible.
 func (b *LabelsBuilder) LabelsResult() LabelsResult {
 	// unchanged path.
-	if len(b.del) == 0 && len(b.add) == 0 && b.err == "" {
+	if !b.hasDel() && !b.hasAdd() && !b.HasErr() {
 		return b.currentResult
 	}
-	return b.toResult(b.labels())
+
+	stream := b.labels(StreamLabel).Copy()
+	structuredMetadata := b.labels(StructuredMetadataLabel).Copy()
+	parsed := b.labels(ParsedLabel).Copy()
+	b.buf = flattenLabels(b.buf, stream, structuredMetadata, parsed)
+	hash := b.hasher.Hash(b.buf)
+	if cached, ok := b.resultCache[hash]; ok {
+		return cached
+	}
+
+	result := NewLabelsResult(b.buf.String(), hash, stream, structuredMetadata, parsed)
+	b.resultCache[hash] = result
+
+	return result
 }
 
-func (b *BaseLabelsBuilder) toResult(buf labels.Labels) LabelsResult {
+func flattenLabels(buf labels.Labels, many ...labels.Labels) labels.Labels {
+	var size int
+	for _, lbls := range many {
+		size += len(lbls)
+	}
+
+	if buf == nil || cap(buf) < size {
+		buf = make(labels.Labels, 0, size)
+	} else {
+		buf = buf[:0]
+	}
+
+	for _, lbls := range many {
+		buf = append(buf, lbls...)
+	}
+	sort.Sort(buf)
+	return buf
+}
+
+func (b *BaseLabelsBuilder) toUncategorizedResult(buf labels.Labels) LabelsResult {
 	hash := b.hasher.Hash(buf)
 	if cached, ok := b.resultCache[hash]; ok {
 		return cached
 	}
-	res := NewLabelsResult(buf.Copy(), hash)
+
+	res := NewLabelsResult(buf.String(), hash, buf, nil, nil)
```

Comment:
I think we should be doing a copy of the buf since it is reused?

Reply:
Good catch, done